### PR TITLE
PRIME-2389 add option to select health authority for OBO Job site

### PR DESCRIPTION
**Pull request by bergomi02**
https://pr-2281.pharmanetenrolment.gov.bc.ca/info
I have created the renewal enrolment test case with PRIMET001 in PR environment.

Summary of the changes:
- new warning in overview page for OBO Job site that the health authority is missing.
image

- if the OBO job site page is missing a health authority, a dropdown will show up. 
image

- a validation is added to ensure each selected health authority must have one OBO job site entry. 
image

Note: the logic for existing enrolment around the health authority and the OBO Job Site did not change.

### CODE REVIEW

**Comment by neophyte57** [comment]
After creating a No College Licence enrollment, signing TOA, etc., I manipulated the data to mimic the problem in PROD, i.e. remove `EnrolleeHealthAuthority` row, and make `OboSite.HealthAuthorityCode` NULL.  However, running into problem when updating enrollment:
1. On "Enrolment Information" screen (Overview), I see the appropriate error messages
2. Edit "Care Setting", select Fraser Health, Continue
3. See the following ... correct

image

4. Edit "Job Site Information", see 

image

5. Click Back
6. See the following, i.e. HA selection removed? 

image

**Comment by bergomi02** [comment]
> After creating a No College Licence enrollment, signing TOA, etc., I manipulated the data to mimic the problem in PROD, i.e. remove `EnrolleeHealthAuthority` row, and make `OboSite.HealthAuthorityCode` NULL. However, running into problem when updating enrollment:
> 
> 1. On "Enrolment Information" screen (Overview), I see the appropriate error messages
> 2. Edit "Care Setting", select Fraser Health, Continue
> 3. See the following ... correct
> 
> image
> 
> 4. Edit "Job Site Information", see
> 
> image
> 
> 5. Click Back
> 6. See the following, i.e. HA selection removed?
> 
> image

This should be fixed.

**Comment by neophyte57** [comment]
A reported situation is Enrollee ID = 607 (Care Setting = HA, No EnrolleeHealthAuthority, No Certification, No OboSite).  In Production or `develop`, after user selects specific HA, when they get to Job Site, they see 

image

Your changes handle this much better but if user goes Back before Saving Job Site details, they can end up in situation as per screenshot.  Can this be handled better?  That said, user could Reload page and then revisit Job Site page and see expected fields.

**Comment by bergomi02** [comment]
> A reported situation is Enrollee ID = 607 (Care Setting = HA, No EnrolleeHealthAuthority, No Certification, No OboSite). In Production or `develop`, after user selects specific HA, when they get to Job Site, they see
> 
> image
> 
> Your changes handle this much better but if user goes Back before Saving Job Site details, they can end up in situation as per screenshot. Can this be handled better? That said, user could Reload page and then revisit Job Site page and see expected fields.

I found the issue but I need more time to fix it. Probably move this to Production first and push the fix of this issue later.

**Comment by bergomi02** [comment]
> A reported situation is Enrollee ID = 607 (Care Setting = HA, No EnrolleeHealthAuthority, No Certification, No OboSite). In Production or `develop`, after user selects specific HA, when they get to Job Site, they see
> 
> image
> 
> Your changes handle this much better but if user goes Back before Saving Job Site details, they can end up in situation as per screenshot. Can this be handled better? That said, user could Reload page and then revisit Job Site page and see expected fields.

Actually, I found a easy fix. I was on the wrong path looking for solution. Anyhow, it should be fixed now.

**Comment by bergomi02** [comment]
@neophyte57 
The validation issue has been fixed. Also, I update the logic when to add a new HA obo site with the option to change HA. It only does it when it detects at least one existing HA obo site without a HA selected. if the existing HA obo sites have their HA selected, the additional HA obo site will be added with the additional HA pre-defined. Hope that makes sense.

**Comment by neophyte57** [comment]
These changes are addressing a number of existing production issues 👍.  One is PRIME-2251 (e.g. Enrollee ID = 21106).  While the improvements allow the user to change their Care Setting from HA to PCHP, the **previous** PCHP `OboSite` is not removed from database.

**Comment by neophyte57** [comment]
> This may be expanding the scope of this PR but if it can be addressed easily, that would be great: e.g. Enrollee ID = 13333, Care Setting = Community Pharmacy, 2 Job Sites -> wants to add HA but can't add Job Site for HA due to missing fields for selected HA

Never mind, looks fixed with latest commit!  Super 💯

**Comment by bergomi02** [comment]
> These changes are addressing a number of existing production issues 👍. One is PRIME-2251 (e.g. Enrollee ID = 21106). While the improvements allow the user to change their Care Setting from HA to PCHP, the **previous** PCHP `OboSite` is not removed from database.

Finally, I think I got it fixed.

File: prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts [hunk 1]
```diff
@@ -171,48 +182,89 @@ export class OboSitesPageComponent extends BaseEnrolmentProfilePage implements O
...
+          if (!this.deviceProviderSites.length) {
+            this.addOboSite(careSettingCode);
           }
-          case CareSettingEnum.HEALTH_AUTHORITY: {
+          break;
+        }
+        case CareSettingEnum.HEALTH_AUTHORITY: {
+          //determine if necessary to add HA job site
+          if (this.oboSites?.length) {
+            let haSiteNum = this.oboSites?.value.filter((s) => s.careSettingCode === 1).length;
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts:239]: Should reference `CareSettingEnum`

File: prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts [hunk 2]
```diff
@@ -171,48 +182,89 @@ export class OboSitesPageComponent extends BaseEnrolmentProfilePage implements O
...
+            if (haSiteNum < this.enrolleeHealthAuthorities.length) {
+              let hasSiteWOHa = false;
```
[the remark this answers is not in the document]
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts]: Good suggestion, it is fixed.

File: prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts [hunk 3]
```diff
@@ -98,6 +101,12 @@ export class OboSitesPageComponent extends BaseEnrolmentProfilePage implements O
     return careSettings;
   }
 
+  public get enrolleeHealthAuthorities() {
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts:104]: Return type would be helpful?

File: prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts [hunk 4]
```diff
@@ -305,4 +357,29 @@ export class OboSitesPageComponent extends BaseEnrolmentProfilePage implements O
       sitesOfHealthAuthority.clear();
     });
   }
+
+  /**
+   * Check if each selected HA has at least one obo job site
+   * @returns true if pass
+   */
+  private validateHAOboSiteJob(): boolean {
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts:365]: Better name?  `eachSelectedHAHasOboSite`?

File: prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts [hunk 5]
```diff
@@ -373,28 +373,37 @@ export class EnrolmentFormStateService extends AbstractFormStateService<Enrolmen
...
-          break;
-        }
-        case CareSettingEnum.HEALTH_AUTHORITY: {
-          this.addHealthAuthorityOboSite(site, healthAuthoritySites, s.healthAuthorityCode);
-          break;
+      const careSetting = this.careSettingsForm.get('careSettings');
+
+      if (careSetting && (careSetting.value.length === 0 || careSetting.value.filter((c) => c.careSettingCode === s.careSettingCode).length > 0)) {
+
+        const site = this.buildOboSiteForm();
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts:380]: Can you make that `siteForm` and add a type to it?  Hmm, I realize how it was before

Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts:380]: it is fixed

File: prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts [hunk 6]
```diff
@@ -373,28 +373,37 @@ export class EnrolmentFormStateService extends AbstractFormStateService<Enrolmen
...
-          break;
-        }
-        case CareSettingEnum.DEVICE_PROVIDER: {
-          this.addNonHealthAuthorityOboSite(site, deviceProviderSites);
-          break;
-        }
-        case CareSettingEnum.HEALTH_AUTHORITY: {
-          this.addHealthAuthorityOboSite(site, healthAuthoritySites, s.healthAuthorityCode);
-          break;
+      const careSetting = this.careSettingsForm.get('careSettings');
```
[the remark this answers is not in the document]
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts]: it is fixed

File: prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts [hunk 7]
```diff
@@ -373,28 +373,37 @@ export class EnrolmentFormStateService extends AbstractFormStateService<Enrolmen
...
+
+      if (careSetting && (careSetting.value.length === 0 || careSetting.value.filter((c) => c.careSettingCode === s.careSettingCode).length > 0)) {
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts]: I'm not sure why this condition is needed?  Can you explain with a comment?

File: prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts (hunk 3 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/obo-sites-page/obo-sites-page.component.ts]: it is fixed

File: prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts (hunk 5 above)
Comment (neophyte57) [reply prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts:380]: Was this updated?  Don't see it

Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts:380]: image
The comment is added

Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/shared/services/enrolment-form-state.service.ts]: Ops, the type is added now.

**Review by neophyte57 — APPROVED** [review]
Amazing work, this addresses so many problems!

**Pull request merged**